dra: reserve + publish during pod binding

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This moves adding a pod to ReservedFor out of the main scheduling cycle into
PreBind. There it is done concurrently in different goroutines. For claims
which were specifically allocated for a pod (the most common case), that
usually makes no difference because the claim is already reserved.
    
It starts to matter when that pod then cannot be scheduled for other reasons,
because then the claim gets unreserved to allow deallocating it. It also
matters for claims that are created separately and then get used multiple times
by different pods.
    
Because multiple pods might get added to the same claim rapidly independently
from each other, it makes sense to do all claim status updates via patching:
then it is no longer necessary to have an up-to-date copy of the claim because
the patch operation will succeed if (and only if) the patched claim is valid.

The same delayed update in PreBind is also used for PodSchedulingContext.

#### Which issue(s) this PR fixes:

Related-to https://github.com/kubernetes/kubernetes/issues/113700

#### Special notes for your reviewer:

Depends on some other pending PRs:
- [x] https://github.com/kubernetes/kubernetes/pull/122276
- [x] https://github.com/kubernetes/kubernetes/pull/121877

#### Does this PR introduce a user-facing change?
```release-note
When scheduling a mixture of pods using ResourceClaims and others which don't, scheduling a pod with ResourceClaims impacts scheduling latency less.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3063
```

pohly committed:
scheduler: publish PodSchedulingContext during PreBind

Blocking API calls during a scheduling cycle like the DRA plugin is doing slow
down overall scheduling, i.e. also affecting pods which don't use DRA.

It is easy to move the blocking calls into a goroutine while the scheduling
cycle ends with "pod unschedulable". The hard part is handling an error when
those API calls then fail in the background. There is a solution for that
(see https://github.com/kubernetes/kubernetes/pull/120963), but it's complex.

Instead, publishing the modified PodSchedulingContext can also be done
later. In the more common case of a pod which is ready for binding except for
its claims, that'll be in PreBind, which runs in a separate goroutine already.

In the less common case that a pod cannot be scheduled, that'll be in
Unreserve which is still blocking.

pohly committed:
dra: patch ReservedFor during PreBind

This moves adding a pod to ReservedFor out of the main scheduling cycle into
PreBind. There it is done concurrently in different goroutines. For claims
which were specifically allocated for a pod (the most common case), that
usually makes no difference because the claim is already reserved.

It starts to matter when that pod then cannot be scheduled for other reasons,
because then the claim gets unreserved to allow deallocating it. It also
matters for claims that are created separately and then get used multiple times
by different pods.

Because multiple pods might get added to the same claim rapidly independently
from each other, it makes sense to do all claim status updates via patching:
then it is no longer necessary to have an up-to-date copy of the claim because
the patch operation will succeed if (and only if) the patched claim is valid.

Server-side-apply cannot be used for this because a client always has to send
the full list of all entries that it wants to be set, i.e. it cannot add one
entry unless it knows the full list.

pohly commented:
This has no significant performance impact for the test cases with just pods that use DRA:
```
name                                                                      old SchedulingThroughput/Average                             new SchedulingThroughput/Average                             delta
PerfScheduling/SchedulingWithResourceClaimTemplate/2000pods_100nodes-36                                                     28.1 ±23%                                                    28.3 ±14%     ~     (p=0.730 n=5+5)
PerfScheduling/SchedulingWithResourceClaimTemplate/2000pods_200nodes-36                                                     45.9 ± 5%                                                    46.4 ± 2%     ~     (p=0.905 n=5+5)
PerfScheduling/SchedulingWithMultipleResourceClaims/2000pods_100nodes-36                                                    17.0 ±13%                                                    18.0 ±39%     ~     (p=1.000 n=5+5)
PerfScheduling/SchedulingWithMultipleResourceClaims/2000pods_200nodes-36                                                    26.4 ±37%                                                    30.3 ±16%     ~     (p=0.310 n=5+5)
```

A new test case specifically for mixed scheduling would have to be added to demonstrate the expected effect.

bg-furiosa commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1040,29 +1006,91 @@ func (pl *dynamicResources) Unreserve(ctx context.Context, cs *framework.CycleSt
 	}
 }
 
+// PreBind gets called in a separate goroutine after it has been determined
```
Quick link to scheduler framework impl
https://github.com/kubernetes/kubernetes/blob/45656f8774c7c761f87a593f68563ccdb3b67523/pkg/scheduler/schedule_one.go#L118

bg-furiosa commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1040,29 +1006,91 @@ func (pl *dynamicResources) Unreserve(ctx context.Context, cs *framework.CycleSt
-	for index, claim := range state.claims {
+
+	// Was publishing delayed? If yes, do it now.
+	if state.podSchedulingState.isDirty() {
+		if err := state.podSchedulingState.publish(ctx, pod, pl.clientset); err != nil {
+			logger.Error(err, "publish PodSchedulingContext")
+		}
+	}
```
Should we publish here?
The function `podSchedulingState.isDirty` does not check delta and simply return true if some fields are populated.
So publish will be called.
1.  Even though it was not called due to the failure of Reserve plugins.
1.  Even though it was called in the previous step: PreBind.

In the first scenario, we already know a node is not suitable, could we get rid of the node from selected node or potential nodes? since It would trigger driver to call `UnsutableNodes` and `allocateClaims`. Maybe calling out  `UnsutableNodes` is necessary but we may no need to push driver call `allocateClaims`since we already know we don't need to do that.

pohly replied:
> So publish will be called.

Yes, but only if there really are pending changes. The fields will be unset if nothing needs to be changed. The code that updates `podSchedulingState` is careful to only set fields which really need to be published.

> Even though it was called in the previous step: PreBind.

`publish` clears fields when it is done, so publish will not get called again.

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1040,29 +1006,91 @@ func (pl *dynamicResources) Unreserve(ctx context.Context, cs *framework.CycleSt
+			// Remove pod from ReservedFor. A strategic-merge-patch is used
+			// because that allows removing an individual entry without having
+			// the latest slice.
+			patch := fmt.Sprintf(`{"metadata": {"uid": %q}, "status": { "reservedFor": [ {"$patch": "delete", "uid": %q} ] }}`,
```
When marshaling an existing object it is impossible to add the special `"$patch": "delete"`. In other words, one can add or overwrite data, but not remove it. That is what we need here.

pohly commented:
This PR is no longer WIP because the prerequisite (support for patching `reservedFor`) was merged.

Let's merge it now to avoid conflicts with the upcoming changes for numeric parameters.

bart0sh commented:
/remove-lifecycle rotten
/triage accepted
/priority important-longterm

pohly commented:
> /priority important-longterm

"long term" = the next few weeks :grinning:

pohly commented:
@bg-furiosa: did I address your questions? Please take another look.

Huang-Wei commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -923,27 +906,11 @@ func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleStat
 	logger := klog.FromContext(ctx)
 	for index, claim := range state.claims {
 		if claim.Status.Allocation != nil {
```
The if clause now wraps an empty block, prefer:

```go
if claim.Status.Allocation != nil {
  continue
}
// Must be delayed allocation.
numDelayedAllocationPending++
```

pohly replied:
Ack.

I didn't want to re-indent the entire else block, that would have obscured the small change that is needed there (the removal of `state.podSchedulingState.publish`). But as you have now seen the change, I can reformat as you suggested :grin:

pohly replied:
Actually, the else clause doesn't change at all here, the `state.podSchedulingState.publish` is further down. Re-indenting still makes sense.

pohly replied:
@Huang-Wei: [update](https://github.com/kubernetes/kubernetes/compare/98a0c9a199296019af8f5d096a403ab90b316e74..6cf42037512b266d9581a745461a131ddfe0da60) pushed, please take another look.

Huang-Wei commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1037,29 +1003,91 @@ func (pl *dynamicResources) Unreserve(ctx context.Context, cs *framework.CycleSt
+			// SSA here to add the pod because then we would have to send the entire slice
+			// or use different field manager strings for each entry.
+			//
+			// With a strategic-merge-patch, we can simply send one new entry. The apiserver
+			// validation will catch if two goroutines try to do that at the same time and
+			// the claim cannot be shared.
```
It's awesome! Thanks.

Huang-Wei commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -803,7 +786,7 @@ func (pl *dynamicResources) PostFilter(ctx context.Context, cs *framework.CycleS
 			claim.Status.ReservedFor = nil
 			logger.V(5).Info("Requesting deallocation of ResourceClaim", "pod", klog.KObj(pod), "resourceclaim", klog.KObj(claim))
-			if err := state.updateClaimStatus(ctx, pl.clientset, index, claim); err != nil {
+			if _, err := pl.clientset.ResourceV1alpha2().ResourceClaims(claim.Namespace).UpdateStatus(ctx, claim, metav1.UpdateOptions{}); err != nil {
```
I have a different option on the usage of PostFilter here, but let's track it in another thread on Slack.

bg-furiosa replied:
https://kubernetes.slack.com/archives/C09TP78DV/p1706170190493659

pohly commented on pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go:
```diff
@@ -1037,29 +1003,91 @@ func (pl *dynamicResources) Unreserve(ctx context.Context, cs *framework.CycleSt
 	logger := klog.FromContext(ctx)
-	for index, claim := range state.claims {
+
+	// Was publishing delayed? If yes, do it now.
+	if state.podSchedulingState.isDirty() {
+		if err := state.podSchedulingState.publish(ctx, pod, pl.clientset); err != nil {
+			logger.Error(err, "publish PodSchedulingContext")
+		}
+	}
```
I've added that as a comment in the code.

pohly committed:
dra scheduler: reformat code

By continuing with the next item in the if clause, the else is no longer needed
and indention can be reduced.

Huang-Wei commented:
/retest

Huang-Wei commented:
/lgtm
/approve
